### PR TITLE
Init commit

### PR DESCRIPTION
**Pull request by hshaosf**
# Service CI Azure serverless python function microservice  
Lightweight configuration file-based workflow system to enable continuous integration for services interacting various APIs and Webhooks. 

## `api/status/http`
Query http status of the serverless function.

### Query
Example
```
$ curl https://<host>/api/status/http

{"status": "success", "data": {"message": "200 OK"}}
```

#### Query Parameters ####
`project`: name of project configuration  
`job`: name of job   
`step`: job step to start  
`limit`: limit the number of steps to run

### Query
Example
```
$ curl https://<host>/api/run/jobs?project=sample_proj_config&job=job1&step=1&limit=1

{"status": "success", "data": {...}}
```

## MODS
This microservice run of concept of leveraging individual modules ([`mods`](./mods)) to interact with various external systems. A mod to used in a `step` within a `job`. A simple one step job that uses version 1 of the email mod to send email looks like this:
```
jobs:
  job1:
    steps:
        - uses: email/send@v1
```

### List of MODs available
* `airtable@v1`: Connects to [Airtable](https://airtable.com/)
    * `get`: gets an Airtable record
    * `insert`: insert a record into Airtable base
* `email@v1`: Connects to [Email microservice](https://github.com/SFDigitalServices/email-microservice-py)
    * `send`: send an email
* `formio@v1`: Connects to [Form.io](https://www.form.io/)
    * `submission-get`: get a form submission
* `jsonata@v1`: Connects to [JSONata microservice](https://github.com/SFDigitalServices/jsonata-fn-js)
    * `eval`: evaluates JSONata query [[try](http://try.jsonata.org/)]
* `pts@v1`: Connects to [Permit Tracking System (PTS) microservice](https://github.com/SFDigitalServices/pts-microservice-fn-py)
    * `permit-create`: creates a building permit application

### CODE REVIEW

File: mods/email_v1.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        url = f"{endpoint}"
+
+        headers = {
+            "x-apikey": os.environ.get('EMAIL_SVC_KEY')
+        }
+
+        print(url)
+        print(data)
+        print(content)
+        response = requests.post(url, headers=headers, data=json.dumps(content), timeout=5)
```
Comment (henryjiang-sfgov) [inline mods/email_v1.py:19]: Maybe move this to a try-catch block? To catch unforeseen errors such as connection time-out.

File: README.md [hunk 2]
```diff
@@ -12,6 +13,42 @@ $ curl https://<host>/api/status/http
...
+#### Query Parameters ####
+`project`: name of project configuration  
+`job`: name of job   
+`step`: job step to start  
+`limit`: limit the number of steps to run
+
+### Query
+Example
+```
+$ curl https://<host>/api/run/jobs?project=sample_proj_config&job=job1&step=1&limit=1
```
Comment (henryjiang-sfgov) [inline README.md:26]: What does the `limit` do? It says limit the number of steps to run,  but there is already a `step` query parameter.

File: run_jobs/__init__.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                                    step_input = out = Step.run(data_step, step_input)
+
+                return func_jsend_response(out, headers=headers, status_code=200)
+
+        raise ValueError("Missing Project")
+
+    #pylint: disable=broad-except
+    except Exception as err:
+        logging.error("Run Step error occurred: %s", traceback.format_exc())
+        msg_error = f"{err}"
```
Comment (henryjiang-sfgov) [inline run_jobs/__init__.py:75]: Merge this to the return statement, like:
`return func_jsend_response(f"{err}", status_code=500)`

File: run_jobs/step.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @staticmethod
+    def run(data: dict, content: dict):
+        """ run step """
+        uses_token = data['uses'].split('/')
+        mod_name = uses_token[0]
+        method_token = uses_token[1].split('@v')
+        method = method_token[0]
+        version = int(method_token[1])
+        print(f"Step.run {mod_name}")
+        mod = importlib.import_module(f".{mod_name}_v{version}", "mods")
```
Comment (henryjiang-sfgov) [inline run_jobs/step.py:17]: Need to check if version exists, import would failed if version is not supported.

File: shared_code/common.py [hunk 5]
```diff
@@ -1,22 +1,40 @@
...
+    return func_jsend_response(
+            func_response, headers=headers, status_code=func_status_code)
+
+def func_jsend_response(body, headers=None, status_code=200):
+    """ func_jsend_success_response """
+    if 200 <= status_code < 300:
+        func_response = json.dumps(jsend.success(body))
+    elif 400 <= status_code < 500:
+        func_response = json.dumps(jsend.fail(body))
+    else:
```
Comment (henryjiang-sfgov) [inline shared_code/common.py:26]: Not sure if you want to be inclusive about other [http_status_code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). I think should be fine just to treat other status as a general error.

File: run_jobs/__init__.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    """ main function for run/step """
+    logging.info('Run/Step processed a request.')
+
+    try:
+        validate_access(req)
+        headers = {
+            "Access-Control-Allow-Origin": "*"
+        }
+
+        if 'project' in req.params:
```
Comment (henryjiang-sfgov) [inline run_jobs/__init__.py:29]: This is up for refactoring: too many nested ifs?  Can't really process a project without other parameters, can we  check project, job, and step all in one if?

File: README.md (hunk 2 above)
Comment (hshaosf) [reply README.md:26]: `limit`: limit the number of steps to run

example: if there are 10 steps, and say we want to only run steps 1-5, then start step=1 and limit =5. 

maybe there is a better way to explain it? open to suggestions :)

File: run_jobs/__init__.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                config_yml = file.read().decode('utf-8')
+                config_json = yaml.load(config_yml, Loader=yaml.SafeLoader)
+                out = data = config_json
+
+                if 'job' in req.params and 'jobs' in data \
+                    and req.params['job'] in data['jobs']:
+
+                    out = data_job = data['jobs'][req.params['job']]
+
+                    if 'step' in req.params and 'steps' in data_job:
```
Comment (josh-chou) [inline run_jobs/__init__.py:43]: if you don't pass in step in the params, then nothing will run?
shouldn't the default be to run all the steps?

Comment (hshaosf) [reply run_jobs/__init__.py]: good call 👍  [d394f66](https://github.com/SFDigitalServices/service-ci-fn-py/pull/1/commits/d394f66c03882a7aa156494d909ecc6298d1acae)

File: run_jobs/__init__.py (hunk 6 above)
Comment (hshaosf) [reply run_jobs/__init__.py]: @henryjiang-sfgov i now moved the job param check up earlier in the if/else statement. we still need to load the config to validate the job param, but now we don't do anything unless they provide both the `project` and `job` param. [9f9c291](https://github.com/SFDigitalServices/service-ci-fn-py/pull/1/commits/9f9c2915dc410a57df252d62bfe3b2ef6cdae02d)

and @josh-chou mentioned that we should default the `step` and i agreed, so we no longer need to checking that [d394f66](https://github.com/SFDigitalServices/service-ci-fn-py/pull/1/commits/d394f66c03882a7aa156494d909ecc6298d1acae)

File: README.md (hunk 2 above)
Comment (henryjiang-sfgov) [reply README.md:26]: ah ok, so it's a range of steps to run, perhaps `start_step` and `end_step`?

File: run_jobs/__init__.py (hunk 3 above)
Comment (hshaosf) [reply run_jobs/__init__.py]: [fba488c](https://github.com/SFDigitalServices/service-ci-fn-py/pull/1/commits/fba488c420fa48c0d499b78189dcec4c6829a127)

**Review by josh-chou — APPROVED** [review]
👍

**Pull request merged**